driver-sync/src/test/functional/com/mongodb/client/unified/UnifiedTest.java

```diff
@@ -414,8 +420,36 @@ private void compareEvents(final UnifiedTestContext context, final BsonDocument
                 context.getEventMatcher().assertConnectionPoolEventsEquality(client, ignoreExtraEvents, expectedEvents,
                         listener.getEvents());
             } else if (eventType.equals("sdam")) {
-                TestServerMonitorListener listener = entities.getServerMonitorListener(client);
-                context.getEventMatcher().assertServerMonitorEventsEquality(client, ignoreExtraEvents, expectedEvents, listener.getEvents());
+
+                // Sdam tests also include topology events, so we need to separate them
+                BsonArray expectedTopologyEvents = new BsonArray(expectedEvents.stream()
+                        .map(BsonValue::asDocument)
+                        .filter(doc -> TOPOLOGY_EVENT_NAMES.stream().anyMatch(doc::containsKey))
+                        .collect(Collectors.toList()));
+
+                if (!expectedTopologyEvents.isEmpty()) {
+                    TestClusterListener clusterListener = entities.getClusterListener(client);
+                    // Unfortunately, some tests expect the cluster to be closed, but do not define it as a waitForEvent in the spec -
+                    // causing a race condition in the test.
+                    if (expectedTopologyEvents.stream().anyMatch(doc -> doc.asDocument().containsKey(TOPOLOGY_CLOSED_EVENT))) {
+                        context.getEventMatcher().waitForClusterClosedEvent(client, clusterListener);
+                    }
+
+                    List<Object> topologyEvents = new ArrayList<>();
+                    topologyEvents.add(clusterListener.getClusterOpeningEvent());
+                    topologyEvents.addAll(clusterListener.getClusterDescriptionChangedEvents());
+                    topologyEvents.add(clusterListener.getClusterClosingEvent());
+                    context.getEventMatcher().assertTopologyEventsEquality(client, ignoreExtraEvents, expectedTopologyEvents, topologyEvents);
+                }
+
+                BsonArray expectedSdamEvents = new BsonArray(expectedEvents.stream()
+                        .map(BsonValue::asDocument)
+                        .filter(doc -> TOPOLOGY_EVENT_NAMES.stream().noneMatch(doc::containsKey))
+                        .collect(Collectors.toList()));
+                if (!expectedSdamEvents.isEmpty()) {
+                    TestServerMonitorListener serverMonitorListener = entities.getServerMonitorListener(client);
+                    context.getEventMatcher().assertServerMonitorEventsEquality(client, ignoreExtraEvents, expectedSdamEvents, serverMonitorListener.getEvents());
+                }
```

Comment:
optional and feel free to ignore and resolve. 

Using the following stream usage maybe improve a little bit, though not too much:
```
Map<Boolean, List<BsonDocument>> partitions = expectedEvents.stream()
                        .map(BsonValue::asDocument)
                        .collect(Collectors.partitioningBy(doc -> TOPOLOGY_EVENT_NAMES.stream().anyMatch(doc::containsKey)));

BsonArray expectedTopologyEvents = new BsonArray(partitions.get(true));
... ...
BsonArray expectedSdamEvents = new BsonArray(partitions.get(false));
... ...
                
                
```

Reply:
Neat - done.